### ISSUE TITLE
Rule proposal: require type annotation on non-initialised variables

### ISSUE THREAD

**Issue by ROpdebee** [typescript-eslint/typescript-eslint#4342]
### Description

We sometimes declare variables without initialising them, to then later initialise them in nested blocks (see the examples below). Those variables are then typed as `any`, which isn't ideal. Although TS' type narrowing does kick in after the assignment, I'd like to see a way to enforce specifying the type of the variable when it is declared to make sure all assignments use the correct types and that TS doesn't wrongly infer unions. It's also better to be more explicit in those situations, IMO.

We're currently using `@typescript-eslint/init-declarations` to show any non-initialised variable declarations, after which we manually check whether a type is specified. If there is, we explicitly disable the rule for that specific line. It's an ugly hack and not fool-proof though.

This could be added as an option to `@typescript-eslint/init-declarations` but at the same time, I feel like that rule exists specifically to forbid any non-initialised variables. It's worth noting that (recent versions of) TypeScript can report usages of variables where they may not yet be defined.

### Fail

```ts
function example1(someInput: string): void {
    let url;
    try {
        url = new URL(someInput);
    } catch {
        console.error(`Invalid URL: ${someInput}`);
        return;
    }

    // Do more with url
    console.log(url.pathname); // url is `URL` here.
}

function example2(someInput: string): void {
    let url;
    try {
        url = new URL(someInput);
    } catch {
        console.error(`Invalid URL: ${someInput}`);
        // return; // Oops, forgot to do an early return
    }

    // Do more with url
    console.log(url.pathname); // TS complains that "object is possibly undefined", which, although true, isn't the root cause.
}

function example3(someInput: string): void {
    let x;
    if (someInput) {
        x = someInput;
    } else {
        x = 42; // TS doesn't complain about this.
    }
   
    console.log(x);  // x is `string | number` here, but we intended it to be just `string`.
}
```

### Pass

```ts
function example1(someInput: string): void {
    let url: URL;
    try {
        url = new URL(someInput);
    } catch {
        console.error(`Invalid URL: ${someInput}`);
        return;
    }

    // Do more with url
    console.log(url.pathname);
}

function example2(someInput: string): void {
    let url: URL;
    try {
        url = new URL(someInput);
    } catch {
        console.error(`Invalid URL: ${someInput}`);
        // return; // Oops, forgot to do an early return
    }

    // Do more with url
    console.log(url.pathname); // TS now complains that url may not be initialised yet, which is the root cause of the issue.
}

function example3(someInput: string): void {
    let x: string;
    if (someInput) {
        x = someInput;
    } else {
        x = 42;  // TS now complain about this line.
    }

    console.log(x);
}
```

**Comment by bradzacher** [comment]
https://typescript-eslint.io/docs/linting/troubleshooting#how-can-i-ban-specific-language-feature

[example playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&showAST=es&code=PQKgUApgzgNglgOwC4AIEHsC0AnaTtwDGSEAJplAJ7ICGAHgFwoDaA5BNtutqwDQoBvMChEooEGBGLcmrAGo0CNAEaSAIlJiKaSbs0RxUAXjQBXGDAC6KAHwoAkqQjI4AMzgcAdHFIMMSAApmJEoABwgAQQR-HTh0BEsASj5hUQBbaCgaAHMIWXR0UKhWMABfSzAQYDAwV1MEYjiEFAg6GjTQyQBGAKh0DPsEUNMkJih8RGzEpgA3dB9BVJFJVFNsGABuJZR8SkXRA5Q1mBQTBAgAdxQAVQAlABle-ohB4aRErYPSlEIdQgALfaHESEeJ9SSeDhcbABAAGgxmNHgpBuDyYABIBH0BkMRqVYR9tgdcEg1ghPqJSjUDsBgCg1OgUGluBAUBdDIDjttQQhwRBPDB0NkAsdPKEdP8EO0IB8ULSjusUHAoChYXd7rCUP8OPyyjU6g0kE0Wm0OpIAExPHFvMYTBBTWbzFFCA4rBWbba7IHA46nNCXVGPbEvXHvCkib6-JAA72HHl8yGcbhwhFIhbqjFY56vPEE8OHeUkskbOV0gDyhSg-Fc3Gy6FQuhQpEZNGaEEUMD2RewCG2VO28oZTJZbI57u5YPQEMFwtF4qQkulsvlABUAMo-fqdGiIFULnQoABE6GUACspKhlShQugoFA4Ko9vUnO5zqRD-wLv8iP9+EiF+gpjZIC+CmBA-DKggrANtqKBcPWPw0KY4ieHqtT1I08Qmu0nQQAAzFaIY2mIdoOigcwLC6ohunQ+ZuCghE5u8sYHHQfrBkx+bfBI4gsaIbEmAALOaJarhuzbQFBqCgmaO7NCogEwcqqFfEsE68lO-IzgEdCyqWKBsVesLjAQ9ooAAPmYaTKBwmrarg-DKCMbKsogJAIE4KKGDsjI2Sgp7IagxmkbCqH9gamFtqauFdD0HGhraplkRRzrbG6xxMOq+ZelRPqKmcAbqoxoaEl8iHRoCuVxpOEJQsm8IIIiyKBpm8VvPipXAnBECkj2XHUqIg6MsyuCjgu44HPGmkCkKIrrGKEpShkhLhRhRpYa0OEWpabUjIlkzTORTqxul6yZQ82XYHsVUHL6BVXEVu1hn25UxjdohTbVSYwg1TXpmiKCYk9HX5jSdLduS+kVkU1a1ghjbNigrYtB2XY9WSfYDSIQ3DqN7LjVyk01Vps1zotS6iXS65oOgVwyduu47P8B6+mkNB7P4KB+QYRppuIKKUD1n7fjGV4Lqy8HSUhvHoK4TOuXeYFhfqa3GptZr4QRT37fah0pSdPUGTr2R0XLxVvIkfEiAJYjZqGXEtDAvHvdbfrCSW+nUxgdNbloiBI8oilM1e8DnCplJY5uGnTrNulbKUQA&rules=N4XyA&tsConfig=N4XyA)
